server/src/main/kotlin/fi/oph/ludos/certificate/Certificate.kt

```diff
@@ -0,0 +1,52 @@
+package fi.oph.ludos.certificate
+
+import fi.oph.ludos.Exam
+import fi.oph.ludos.PublishState
+import java.sql.Timestamp
+import javax.validation.constraints.NotBlank
+
+interface Certificate {
+    val exam: Exam
+    @get:NotBlank
+    val nameFi: String
+    @get:NotBlank
+    val contentFi: String
+    val publishState: PublishState
+    @get:NotBlank
+    val fileName: String
+    @get:NotBlank
+    val fileKey: String
+    @get:NotBlank
+    val fileUploadDate: String
+}
+
+data class CertificateDtoIn(
+    val id: Int,
+    override val exam: Exam,
+    override val nameFi: String,
+    override val contentFi: String,
+    override val publishState: PublishState,
+    override val fileName: String,
+    override val fileKey: String,
+    override val fileUploadDate: String,
```

Comment:
Nyt kun kelailen asiaa, niin tää Certificaten 2-vaiheinen lataus (eka filu, vasta sit lomake + tiedostometadata) alkoi tuntua vähän hassulta, mut on täysin mahdollista, että oon vaan missannut jonkin tärkeän sitä tukevan pointin. Voitais jutella tästä, mut oksennan mun ajatukset tähän jo.

2-vaiheisen latauksen hankaluuksista tulee mieleen ainakin pari juttua:

1) roskaa ja mahdollisesti salaisia/yksityisiä tiedostoja vahingossa ämpäriin: 

Esimerkiksi jos käyttäjä täyttelee lomaketta ja valitsee vahingossa jonkin yksityisen tiedoston ensin, niin se on sit meidän pilviämpärissä, eikä käyttäjä edes tajua sitä, jos korvaa tiedoston oikealla ennen lomakkeen lähetystä. Refreshatessa / sivulta poistuessa ämpäriin jää myös turha roska, jos tiedosto oli valittuna.

Roskat voidaan toki siivota säännöllisesti ja automaattisesti pois, mut se maksaa toteutusaikaa ja suorituskykyä ja monimutkaisuutta.

2) Bäkkäri joutuu luottamaan frontin lähettämiin fileKey / fileUploadDate -arvoihin sekä todistusta luodessa että päivittäessä. Jos frontti vaik bugittaa update-operaatiossa, niin voidaan menettää viittaus tiedostoon.


Olisiko sittenkin parempi lähettää tiedosto samassa kutsussa muun metadatan kanssa? Tässä tavassa olis edellisten korjaamisen lisäksi myös se hyvä puoli, että updatessa vanha tiedosto voidaan poistaa S3:sta, jolloin siellä ei ikinä pitäisi olla roskaa.

Teknisesti yks mahis tämän toteuttamiseen ois käyttää create/update-kutsun Content-Typenä `multipart/mixed`, jossa eka osa vois olla `application/json`, joka sisältää kaiken metadatan, ja toinen osa vois olla itse tiedosto oikealla MIME-typellä, esim `application/pdf`. Create-kutsussa tiedosto ois pakollinen, ja update-kutsussa tiedosto olisi mukana vain, jos se muuttuu. 

Toinen mahis ois käyttää kutsujen Content-Typenä `multipart/form-data`, ja laittaa jokaiseen osioon lomakkeen yksi kenttä, jolloin kutsun bodyssä ei ois ollenkaan JSONia mukana. En tiedä, kumpaan näistä frontti ja Spring Boot taipuu paremmin.

Teoriassa vois lähettää tiedoston sisällön myös yhdessä JSON-kentässä vaik base64-enkoodattuna, mutta sit tiedostoa pitäis itse käsitellä ikävästi, ja se on vaikea tehdä tehokkaasti, joten jättäisin tän hätävaihtoehdoksi, jos päädytään 1-vaiheiseen lähetykseen.